src/vcpkg/commands.xdownload.cpp

```diff
@@ -46,26 +55,60 @@ namespace vcpkg::Commands::X_Download
             parse_download_configuration(args.asset_sources_template).value_or_exit(VCPKG_LINE_INFO)};
         path file = fs.absolute(vcpkg::u8path(args.command_arguments[0]), VCPKG_LINE_INFO);
 
-        std::string sha = Strings::ascii_to_lowercase(std::string(args.command_arguments[1]));
-        if (!is_lower_sha512(sha))
+        Optional<std::string> sha;
+        auto sha_it = parsed.settings.find(OPTION_SHA512);
+        if (args.command_arguments.size() > 1)
         {
-            Checks::exit_with_message(
-                VCPKG_LINE_INFO, "Error: SHA512's must be 128 hex characters: '%s'", args.command_arguments[1]);
+            if (sha_it != parsed.settings.end())
+            {
+                Checks::exit_with_message(VCPKG_LINE_INFO, "Error: SHA512 passed as both an argument and as an option. Only pass one of these.");
+            }
+            sha = args.command_arguments[1];
+        }
+        else if (sha_it != parsed.settings.end())
+        {
+            sha = sha_it->second;
+        }
+
+        if (Util::Sets::contains(parsed.switches, OPTION_SKIP_SHA512))
+        {
+            if (sha.has_value())
+            {
+                Checks::exit_with_message(VCPKG_LINE_INFO, "SHA512 passed, but --skip-sha512 was also passed; only do one or the other.");
+            }
+        }
+        else if (!sha.has_value())
+        {
+            Checks::exit_with_message(VCPKG_LINE_INFO, "Required argument --sha512 was not passed.");
+        }
+        else
+        {
+            auto p = sha.get();
+            if (!is_sha512(*p))
+            {
+                Checks::exit_with_message(
+                    VCPKG_LINE_INFO, "Error: SHA512's must be 128 hex characters: '%s'", *p);
+            }
+            Strings::ascii_to_lowercase(p->begin(), p->end());
         }
 
         // Is this a store command?
         if (Util::Sets::contains(parsed.switches, OPTION_STORE))
         {
+            if (!sha.has_value())
+            {
+                Checks::exit_with_message(VCPKG_LINE_INFO, "--store option is invalid without a sha512.");
+            }
+
             auto s = fs.status(file, VCPKG_LINE_INFO);
             if (s.type() != vcpkg::file_type::regular)
             {
                 Checks::exit_with_message(
                     VCPKG_LINE_INFO, "Error: path was not a regular file: %s", vcpkg::u8string(file));
             }
-            auto hash =
-                Strings::ascii_to_lowercase(Hash::get_file_hash(VCPKG_LINE_INFO, fs, file, Hash::Algorithm::Sha512));
-            if (hash != sha) Checks::exit_with_message(VCPKG_LINE_INFO, "Error: file to store does not match hash");
-            download_manager.put_file_to_mirror(fs, file, sha).value_or_exit(VCPKG_LINE_INFO);
+            auto hash = Hash::get_file_hash(VCPKG_LINE_INFO, fs, file, Hash::Algorithm::Sha512);
```

Comment:
Why does this no longer need lowercasing?

Reply:
It never needed lowercasing - `get_file_hash` returns a lowercase value (see the tests)